File: BackgroundServices/StaleDataCleaner.cs
using BirthdayBot.Data;
using NpgsqlTypes;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace BirthdayBot.BackgroundServices
{
    /// <summary>
    /// Automatically removes database information for guilds that have not been accessed in a long time.
    /// </summary>
    class StaleDataCleaner : BackgroundService
    {
        private int _tickCount = 0;

        public StaleDataCleaner(ShardInstance instance) : base(instance) { }

        public override async Task OnTick(CancellationToken token)
        {
            if (++_tickCount % 20 != 0)
            {
                // Do not process on every tick.
                return;
            }

            // Build a list of all values to update
            var updateList = new Dictionary<ulong, List<ulong>>();
            foreach (var g in ShardInstance.DiscordClient.Guilds)
            {
                // Get list of IDs for all users who exist in the database and currently exist in the guild
                var savedUserIds = from cu in await GuildUserConfiguration.LoadAllAsync(g.Id) select cu.UserId;
                var guildUserIds = from gu in g.Users select gu.Id;
                var existingCachedIds = savedUserIds.Intersect(guildUserIds);
                updateList[g.Id] = new List<ulong>(existingCachedIds);
            }

            using var db = await Database.OpenConnectionAsync();

            // Statement for updating last_seen in guilds
            var cUpdateGuild = db.CreateCommand();
            cUpdateGuild.CommandText = $"update {GuildConfiguration.BackingTable} set last_seen = now() "
                + "where guild_id = @Gid";
            var pUpdateG = cUpdateGuild.Parameters.Add("@Gid", NpgsqlDbType.Bigint);
            cUpdateGuild.Prepare();

            // Statement for updating last_seen in guild users
            var cUpdateGuildUser = db.CreateCommand();
            cUpdateGuildUser.CommandText = $"update {GuildUserConfiguration.BackingTable} set last_seen = now() "
                + "where guild_id = @Gid and user_id = @Uid";
            var pUpdateGU_g = cUpdateGuildUser.Parameters.Add("@Gid", NpgsqlDbType.Bigint);
            var pUpdateGU_u = cUpdateGuildUser.Parameters.Add("@Uid", NpgsqlDbType.Bigint);
            cUpdateGuildUser.Prepare();

            // Do actual updates
            int updatedGuilds = 0;
            int updatedUsers = 0;
            foreach (var item in updateList)
            {
                var guild = item.Key;
                var userlist = item.Value;

                pUpdateG.Value = (long)guild;
                updatedGuilds += await cUpdateGuild.ExecuteNonQueryAsync(token);

                pUpdateGU_g.Value = (long)guild;
                foreach (var userid in userlist)
                {
                    pUpdateGU_u.Value = (long)userid;
                    updatedUsers += await cUpdateGuildUser.ExecuteNonQueryAsync(token);
                }
            }
            Log($"Updated last-seen records: {updatedGuilds} guilds, {updatedUsers} users");

            // Delete all old values - expects referencing tables to have 'on delete cascade'
            using var t = db.BeginTransaction();
            int staleGuilds, staleUsers;
            using (var c = db.CreateCommand())
            {
                // Delete data for guilds not seen in 4 weeks
                c.CommandText = $"delete from {GuildConfiguration.BackingTable} where (now() - interval '28 days') > last_seen";
                staleGuilds = c.ExecuteNonQuery();
            }
            using (var c = db.CreateCommand())
            {
                // Delete data for users not seen in 8 weeks
                c.CommandText = $"delete from {GuildUserConfiguration.BackingTable} where (now() - interval '56 days') > last_seen";
                staleUsers = c.ExecuteNonQuery();
            }
            Log($"Will remove {staleGuilds} guilds, {staleUsers} users.");
            t.Commit();
        }
    }
}

File: BirthdayBot/BackgroundServices/StaleDataCleaner.vb
''' <summary>
''' Automatically removes database information for guilds that have not been accessed in a long time.
''' </summary>
Class StaleDataCleaner
    Inherits BackgroundService

    Public Sub New(instance As BirthdayBot)
        MyBase.New(instance)
    End Sub

    Public Overrides Async Function OnTick() As Task
        ' Build a list of all values to update
        Dim updateList As New Dictionary(Of ULong, List(Of ULong))
        For Each gi In BotInstance.GuildCache
            Dim existingUsers As New List(Of ULong)()
            updateList(gi.Key) = existingUsers

            Dim guild = BotInstance.DiscordClient.GetGuild(gi.Key)
            If guild Is Nothing Then Continue For ' Have cache without being in guild. Unlikely, but...

            ' Get IDs of cached users which are currently in the guild
            Dim cachedUserIds = From cu In gi.Value.Users Select cu.UserId
            Dim guildUserIds = From gu In guild.Users Select gu.Id
            Dim existingCachedIds = cachedUserIds.Intersect(guildUserIds)
            existingUsers.AddRange(existingCachedIds)
        Next

        Using db = Await BotInstance.Config.DatabaseSettings.OpenConnectionAsync()
            ' Prepare to update a lot of last-seen values
            Dim cUpdateGuild = db.CreateCommand()
            cUpdateGuild.CommandText = $"update {GuildStateInformation.BackingTable} set last_seen = now() " +
                "where guild_id = @Gid"
            Dim pUpdateG = cUpdateGuild.Parameters.Add("@Gid", NpgsqlTypes.NpgsqlDbType.Bigint)
            cUpdateGuild.Prepare()

            Dim cUpdateGuildUser = db.CreateCommand()
            cUpdateGuildUser.CommandText = $"update {GuildUserSettings.BackingTable} set last_seen = now() " +
                "where guild_id = @Gid and user_id = @Uid"
            Dim pUpdateGU_g = cUpdateGuildUser.Parameters.Add("@Gid", NpgsqlTypes.NpgsqlDbType.Bigint)
            Dim pUpdateGU_u = cUpdateGuildUser.Parameters.Add("@Uid", NpgsqlTypes.NpgsqlDbType.Bigint)
            cUpdateGuildUser.Prepare()

            ' Do actual updates
            For Each item In updateList
                Dim guild = item.Key
                Dim userlist = item.Value

                pUpdateG.Value = CLng(guild)
                cUpdateGuild.ExecuteNonQuery()

                pUpdateGU_g.Value = CLng(guild)
                For Each userid In userlist
                    pUpdateGU_u.Value = CLng(userid)
                    cUpdateGuildUser.ExecuteNonQuery()
                Next
            Next

            ' Delete all old values - expects referencing tables to have 'on delete cascade'
            Using t = db.BeginTransaction()
                Using c = db.CreateCommand()
                    ' Delete data for guilds not seen in 4 weeks
                    c.CommandText = $"delete from {GuildStateInformation.BackingTable} where (now() - interval '28 days') > last_seen"
                    Dim r = c.ExecuteNonQuery()
                    If r <> 0 Then Log($"Removed {r} stale guild(s).")
                End Using
                Using c = db.CreateCommand()
                    ' Delete data for users not seen in 8 weeks
                    c.CommandText = $"delete from {GuildUserSettings.BackingTable} where (now() - interval '56 days') > last_seen"
                    Dim r = c.ExecuteNonQuery()
                    If r <> 0 Then Log($"Removed {r} stale user(s).")
                End Using
            End Using
        End Using
    End Function
End Class
